### DISCUSSION TITLE
[Feature]: Driving an Arduino Nano from the wand board.

### DISCUSSION THREAD

**Discussion by ThePackDoctor** [gpstar81/GPStar-proton-pack#418] [Q&A]
### What would you like to add in terms of software changes?

I'm aware there's a 5v out, but I'm gonna install the Bargraph. I'm gonna run an Arduino Nano in the wand, with a neopixel jewel (7led) and ring (12led)
We're all aware of how limited the space in the wand is, so I was hoping to run it off of the wand board. 
For a Spengler wand, is the Barrel-LED JST used? If not, could it theoretically be made to put out 5v, again, theoretically, powering only once the vent light turns on? 👀 

### Would this request involve any specific hardware?

Wand Barrel LED JST connector.
Otherwise, I don't believe so? 

### Homework Completed

- [X] I did my part!

**Comment by nomakewan** [comment]
This is absolutely not going to happen, so I've converted it to a discussion.

However, perhaps you should think more about _why_ you would want to run a Nano. Is there something the Nano can do that the GPStar wand board's Mega 2560 cannot?

As an example, we have a user who has a full TVG setup on his wand including a completely separate addressable barrel for the slime blower that activates when slime mode is active, then automatically switches back to the regular barrel for the other firing modes. All from the stock GPStar wand board (and updated firmware, of course).

**Reply by ThePackDoctor** [reply]
I have absolutely no clue how I'd do this. All I want is to run one neopixel jewel, with a spinning red pattern (center LED always on, red, outer 6 creating a baby cyclotron look), toggleable with a rotary encoder between the aforementioned red and a solid white (I'm gonna use an RGBW Cool White jewel from adafruit). 
I don't even know how I would connect something like this to the wand board. Is it possible to tap off the 5v out using a little custom JST splitter while the bargraph still has full functionality?

**Reply by nomakewan** [reply]
Apologies for the delay in responding; was on vacation and Github doesn't work on my phone (very cool breaking github on mobile Safari).

An additional neopixel channel can be run off of the vent/top LED pins. Doing this will actually already be supported natively in the next firmware release for the top LED pin, but it's trivial to add another for the vent pin as well with the same line of code (just changing the pin name).

If you're only changing between two colors, is there a reason why you need a rotary encoder? Couldn't it just be an on/off switch? Rotary encoder uses several pins and has complex code, whereas an on/off switch uses only a single pin and needs only two lines of code.

All 5V out on the board are tied together, so you can use any one you want. Same with GNDs.

**Reply by ThePackDoctor** [reply]
I'm just using the button pins for the encoder. I'm doing it this way because I want to make more knobs on my wand able to be interacted with physically, not digitally. In this case, the knob to the side of the clippard. The encoder function itself is completely unused, I actually bent down the pins on the encoder I used, to get them out of the way. It's essentially just a baby push button momentary switch, but fancier :>

**Reply by nomakewan** [reply]
Ah okay. That's better then. As for the switch you can make a rotary motion that has a switch just fine even without a rotary encoder. In fac,t, Hasbro did exactly this with the rotary knob on the front of the ghost trap: it's a knob that internally has a raised and lowered section, and there is a microswitch next to it. When you rotate the knob it feels like it has distinct stopping positions, but regardless of the direction you turn the knob it actually just either turns on or off the switch.

In any case, if you're just using it as a single switch then that's the same thing, and thankfully that means you can use the normal switch code too. This thing you're doing shouldn't be too bad. I imagine the hardest part will just be making that custom animation you want. Implementing it into the GPStar hardware should actually be the easy part.

**Reply by ThePackDoctor** [reply]
I have zero idea how to do this. A point in the right direction to resources would be much appreciated.
tysm for the help so far!!

**Reply by nomakewan** [reply]
What do you mean? You mentioned using a Nano with a Jewel and Ring. Coding for the Nano is almost identical to coding for the Mega 2560. So essentially you'd just do the exact same thing you were already going to do, except instead of writing from scratch for a Nano, you download our source code and code it in the wand INO instead.

If there is something specific you need help with, please clarify so I can try to direct you to the correct resource.

**Reply by ThePackDoctor** [reply]
> What do you mean? You mentioned using a Nano with a Jewel and Ring. Coding for the Nano is almost identical to coding for the Mega 2560. So essentially you'd just do the exact same thing you were already going to do, except instead of writing from scratch for a Nano, you download our source code and code it in the wand INO instead.

Did not know this. (Doesn't help I know very little about programming) Never used neopixels before. I'm gonna read up on Adafruit's guides, but where would I start (say, downloading the source code here for the kit, where would that even be?) sorry if I seem clueless (it's because I am)

**Reply by nomakewan** [reply]
Oh, I thought you already had experience with these and were just looking for a way to do it. I was not aware you had no Arduino and/or NeoPixel programming experience. That might be a bit of a learning curve for you. In that case I would highly recommend that you get yourself an Arduino starter kit of some sort and learn how to use it. And yes, of course Adafruit has excellent and very helpful guides as well.

The source code is available by clicking the green "Code" button on the main repository page, then clicking "Download ZIP".

![getcode](https://github.com/user-attachments/assets/c6700d08-276d-40bb-9055-d1c3f3edf14d)

Then extract it somewhere, and go into the `source\NeutronaWand` folder and you will find all the source code for the Neutrona Wand.

The main file is `NeutronaWand.ino`, and this is where things like setting up the hardware are done. So that's probably where you'll want to insert whatever code you made. But again, since you're starting from zero, I highly recommend that you just try to get whatever it is you're trying to do working with an Arduino starter kit first, and then once you have it doing what you want you should then understand how an Arduino program looks and thus how to put the code you just made into our existing code.

Best of luck--we've all been there. None of us are born with C++ in our veins. As long as you want to learn, you will be able to learn. Enjoy!

**Reply by ThePackDoctor** [reply]
> Oh, I thought you already had experience with these and were just looking for a way to do it. I was not aware you had no Arduino and/or NeoPixel programming experience. That might be a bit of a learning curve for you. In that case I would highly recommend that you get yourself an Arduino starter kit of some sort and learn how to use it. And yes, of course Adafruit has excellent and very helpful guides as well.
> 
> 
> 
> The source code is available by clicking the green "Code" button on the main repository page, then clicking "Download ZIP".
> 
> 
> 
> ![getcode](https://github.com/user-attachments/assets/c6700d08-276d-40bb-9055-d1c3f3edf14d)
> 
> 
> 
> Then extract it somewhere, and go into the `source\NeutronaWand` folder and you will find all the source code for the Neutrona Wand.
> 
> 
> 
> The main file is `NeutronaWand.ino`, and this is where things like setting up the hardware are done. So that's probably where you'll want to insert whatever code you made. But again, since you're starting from zero, I highly recommend that you just try to get whatever it is you're trying to do working with an Arduino starter kit first, and then once you have it doing what you want you should then understand how an Arduino program looks and thus how to put the code you just made into our existing code.
> 
> 
> 
> Best of luck--we've all been there. None of us are born with C++ in our veins. As long as you want to learn, you will be able to learn. Enjoy!

Thank you!! I know some basics of arduino and have a starter kit, but I never went super in depth. This'll be useful!!

**Reply by ThePackDoctor** [reply]
> Best of luck--we've all been there. None of us are born with C++ in our veins. As long as you want to learn, you will be able to learn. Enjoy!

OKAY! I bought the spongeface arduino reboot pack 7 segment displays and learned a dash of C++ and Arduino because of it, and I finally understand the stuff you said. My question is now physical hardware. Would I connect it to the same Neopixel line as the RGB vent light, and if so, what should I use for the white cliplite lens? (also ideally RGB, but I'm okay with it not being such)

**Reply by nomakewan** [reply]
The RGB vent light includes the light for the white cliplite lens. Both the white cliplite lens and the vent light are RGB.

Going forward our intent is that any additional RGB LEDs added to the wand (such as TVG addons) will be chained from the RGB vent light pin, yes.

**Reply by ThePackDoctor** [reply]
> The RGB vent light includes the light for the white cliplite lens. Both the white cliplite lens and the vent light are RGB.

Again, I installed the RGB vent light and am already aware of this fact. If I'm installing the neopixel jewel, I don't think I'll be able to install both the jewel and RGB vent light board, physically. Therefore, my question being "what should I use for the white cliplite lens" pertains to how I am supposed to run that light, physically. 
Should I:
A: try to run the jewel ABOVE the vent light (not a terrible idea, would be very cramped tho, if it even fits at all.) 
B: figure out another solution for the cliplite, using a separate neopixel diode.

**Reply by nomakewan** [reply]
You did not actually state anywhere _where_ these additional neopixel components were going to go, so I had no reason to believe that they would have caused any sort of conflict with an RGB vent light board placement.

If for some reason they do cause a conflict, then you will need to work around that as necessary for the specific space restraints you have with your install. If the RGB vent light board fits, great. If it doesn't, you might need to create a custom PCB that is essentially just half of the RGB vent light board for just the blinking light.

**Reply by ThePackDoctor** [reply]
> You did not actually state anywhere _where_ these additional neopixel components were going to go, so I had no reason to believe that they would have caused any sort of conflict with an RGB vent light board placement.
> 

Could’ve sworn I mentioned it. Weird.

> If for some reason they do cause a conflict, then you will need to work around that as necessary for the specific space restraints you have with your install. If the RGB vent light board fits, great. If it doesn't, you might need to create a custom PCB that is essentially just half of the RGB vent light board for just the blinking light.

Eugh. There’s enough room, I’m pretty sure, it might be close though. I know NOTHING about PCB design and have no interest in spending money on that route. I’ll figure it out.
